# SCARlink

## Running SCARlink

### Step 1. SCARlink preprocessing

SCARlink requires preprocessed scRNA-seq in a Seurat/anndata object and scATAC-seq (non-binary tile matrix) in ArchR/anndata object. The cell names in both the objects need to be identical. See the code in ```|-- code/model/SCARlink/Data_preparation```.

```scarlink_preprocessing``` generates a coassay HDF5 file (```coassay_matrix.h5```) containing gene expression vectors and corresponding tile-accessibility matrices for highly variable genes. The output is saved in a specified directory (e.g., ```multiome_out```), which should be used for all subsequent SCARlink steps.

**Sample Command**:

```
scarlink_processing --scrna pbmc.rds --scatac pbmc_Seurat_to_ArchR -o multiome_out -g hg38
```

### Step 2. Running SCARlink

```scarlink``` computes the gene-level regression models and, when cell type information is provided using ```-c```, estimates cell-type-specific Shapley values. The same output directory (```multiome_out```) generated by ```scarlink_preprocessing``` should be used for all subsequent SCARlink steps.

**Note:** The ```celltype``` must be present in the input Seurat or ArchR object.

**Sample Command**:

```
scarlink -o multiome_out -g hg38 -c celltype
```

### Step 3. Predict gene‚Äêlinked tiles

```scarlink_tiles``` outputs tile-level significance for each gene and cell type. It should be run using the same output directory (```multiome_out```) as the previous SCARlink steps.

**Sample Command**:

```
scarlink_tiles -o multiome_out -c celltype 
```
